File: src/test/java/com/github/dkoval/leetcode/challenge/ChampagneTowerTest.java
package com.github.dkoval.leetcode.challenge;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class ChampagneTowerTest {

    public static List<Arguments> input() {
        return Arrays.asList(
                Arguments.of(1, 1, 1, 0.0),
                Arguments.of(2, 1, 1, 0.5),
                Arguments.of(100000009, 33, 17, 1.0)
        );
    }

    private final ChampagneTower solution = new ChampagneTower();

    @ParameterizedTest
    @MethodSource("input")
    void champagneTower(int poured, int query_row, int query_glass, double expected) {
        double actual = solution.champagneTower(poured, query_row, query_glass);
        assertEquals(expected, actual, 1E-6);
    }
}

File: src/test/kotlin/com/github/dkoval/leetcode/challenge/ChampagneTowerTest.kt
package com.github.dkoval.leetcode.challenge

import com.github.dkoval.leetcode.challenge.ChampagneTower.ChampagneTowerRev1
import com.github.dkoval.leetcode.challenge.ChampagneTower.ChampagneTowerRev2
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.ArgumentsProvider
import org.junit.jupiter.params.provider.ArgumentsSource
import java.util.stream.Stream

internal class ChampagneTowerTest {

    class InputArgumentsProvider : ArgumentsProvider {

        override fun provideArguments(context: ExtensionContext): Stream<out Arguments> = Stream.of(
            Arguments.of(1, 1, 1, 0.0),
            Arguments.of(2, 1, 1, 0.5),
            Arguments.of(1, 1, 0, 0.0),
            Arguments.of(100000009, 33, 17, 1.0)
        )
    }

    @Nested
    inner class ChampagneTowerRev1Test {

        @ParameterizedTest
        @ArgumentsSource(InputArgumentsProvider::class)
        fun `should return how full the jth glass in the ith row is`(
            poured: Int,
            queryRow: Int,
            queryGlass: Int,
            expected: Double
        ) {
            ChampagneTowerRev1().test(poured, queryRow, queryGlass, expected)
        }
    }

    @Nested
    inner class ChampagneTowerRev2Test {

        @ParameterizedTest
        @ArgumentsSource(InputArgumentsProvider::class)
        fun `should return how full the jth glass in the ith row is`(
            poured: Int,
            queryRow: Int,
            queryGlass: Int,
            expected: Double
        ) {
            ChampagneTowerRev2().test(poured, queryRow, queryGlass, expected)
        }
    }
}

private fun ChampagneTower.test(poured: Int, queryRow: Int, queryGlass: Int, expected: Double) {
    val actual = champagneTower(poured, queryRow, queryGlass)
    assertEquals(expected, actual, 1E-6)
}
